File: internal/channel/channel.go
package channel

import (
	"encoding/json"
	"time"
)

// Channel 是 channels 表的 ORM 模型与领域对象。
type Channel struct {
	ID          int64           `gorm:"primaryKey;column:id"         json:"id"`
	Name        string          `gorm:"column:name;size:64"           json:"name"`
	Provider    string          `gorm:"column:provider;size:32"       json:"provider"`
	BaseURL     string          `gorm:"column:base_url;size:256"      json:"base_url"`
	Credentials string          `gorm:"column:credentials;type:text"  json:"-"` // 密文
	Priority    int16           `gorm:"column:priority"               json:"priority"`
	Weight      int             `gorm:"column:weight"                 json:"weight"`
	Status      int8            `gorm:"column:status"                 json:"status"`
	Tags            json.RawMessage `gorm:"column:tags;type:json"             json:"tags"`
	Extra           json.RawMessage `gorm:"column:extra;type:json"            json:"extra"`
	PoolEnabled     int8            `gorm:"column:pool_enabled"               json:"pool_enabled"`
	AccountStrategy string          `gorm:"column:account_strategy;size:16"   json:"account_strategy"`
	AccountTopK     int8            `gorm:"column:account_top_k"              json:"account_top_k"`
	CreatedAt       time.Time       `gorm:"column:created_at"                 json:"created_at"`
	UpdatedAt   time.Time       `gorm:"column:updated_at"             json:"updated_at"`
	DeletedAt   *time.Time      `gorm:"column:deleted_at;index"       json:"deleted_at,omitempty"`

	// 运行时字段(不持久化)
	Cred           Credential               `gorm:"-" json:"-"`
	ModelMap       map[string]string        `gorm:"-" json:"-"`
	ModelOverrides map[string]ModelOverride `gorm:"-" json:"-"`
}

// TableName implements gorm.Tabler.
func (Channel) TableName() string { return "channels" }

// ModelMapping 是 channel_model_mappings 表的 ORM 模型。
type ModelMapping struct {
	ID             int64     `gorm:"primaryKey;column:id"                       json:"id"`
	ChannelID      int64     `gorm:"column:channel_id"                          json:"channel_id"`
	ClientModel    string    `gorm:"column:client_model;size:128"               json:"client_model"`
	UpstreamModel  string    `gorm:"column:upstream_model;size:128"             json:"upstream_model"`
	InputRatio     *float64  `gorm:"column:input_ratio;type:decimal(10,4)"      json:"input_ratio"`
	OutputRatio    *float64  `gorm:"column:output_ratio;type:decimal(10,4)"     json:"output_ratio"`
	CachedRatio    *float64  `gorm:"column:cached_ratio;type:decimal(10,4)"     json:"cached_ratio"`
	ReasoningRatio *float64  `gorm:"column:reasoning_ratio;type:decimal(10,4)"  json:"reasoning_ratio"`
	CreatedAt      time.Time `gorm:"column:created_at"                          json:"created_at"`
	UpdatedAt      time.Time `gorm:"column:updated_at"                          json:"updated_at"`
}

// TableName implements gorm.Tabler.
func (ModelMapping) TableName() string { return "channel_model_mappings" }

// Credential 是渠道凭证的应用层结构。
type Credential struct {
	APIKey  string            `json:"api_key"`
	Secret  string            `json:"secret,omitempty"`
	Region  string            `json:"region,omitempty"`
	BaseURL string            `json:"base_url,omitempty"`
	Extra   map[string]string `json:"extra,omitempty"`
}

// IsZero 判断凭证是否为空。
func (c Credential) IsZero() bool { return c.APIKey == "" }

// ModelOverride 是单渠道模型倍率覆盖。nil 字段用模型字典默认值。
type ModelOverride struct {
	InputRatio     *float64
	OutputRatio    *float64
	CachedRatio    *float64
	ReasoningRatio *float64
}

// SelectHint 是 Selector 的辅助参数。
type SelectHint struct {
	Excluded []int64  // 本次请求已失败的 channel id
	Tags     []string // 要求 channel 必须包含的 tag
	UserID   int64
	TokenID  int64
	GroupID  int64
}

// MaskAPIKey 脱敏 API key:首 3 + **** + 末 4;不足 8 字符全 ****。
func MaskAPIKey(key string) string {
	if len(key) < 8 {
		return "****"
	}
	return key[:3] + "****" + key[len(key)-4:]
}

File: web/admin/src/api/channel.ts
import { get, post, patch, del, type Page } from './http'

export interface Channel {
  id: number; name: string; provider: string
  base_url: string
  priority: number; weight: number
  status: 0 | 1 | 2
  group_id?: number
  tags: string[]
  extra: Record<string, unknown>
  credentials_masked?: { api_key?: string }
  created_at: string; updated_at: string
  health?: { state: 'closed' | 'half_open' | 'open'; consec_fail: number; opened_at: string | null }
  last_test?: { at: string; ok: boolean; latency_ms?: number; error?: string }
}

export interface ChannelListParams {
  page?: number; size?: number; keyword?: string
  provider?: string; status?: 0 | 1 | 2; tags?: string[]; group_id?: number
}

export interface ChannelInput {
  name: string; provider: string; base_url?: string
  credentials?: { secret?: string; region?: string; extra?: Record<string, string> }
  priority: number; weight: number
  status: 0 | 1
  group_id?: number
  tags: string[]; extra: Record<string, unknown>
}

export interface Mapping {
  id?: number; channel_id?: number
  client_model: string; upstream_model: string
  input_ratio?: number | null; output_ratio?: number | null
  cached_ratio?: number | null; reasoning_ratio?: number | null
}

export interface TestResult { ok: boolean; latency_ms?: number; error?: string; channel_id?: number; model?: string }

export const channelApi = {
  list: (p: ChannelListParams) => get<Page<Channel>>('/api/admin/channels', p as Record<string, unknown>),
  create: (b: ChannelInput) => post<Channel>('/api/admin/channels', b),
  detail: (id: number) => get<Channel>(`/api/admin/channels/${id}`),
  patch: (id: number, b: Partial<ChannelInput>) => patch<Channel>(`/api/admin/channels/${id}`, b),
  remove: (id: number) => del<{ ok: true }>(`/api/admin/channels/${id}`),
  enable: (id: number) => post<{ ok: true; status: 0 }>(`/api/admin/channels/${id}/enable`, {}),
  disable: (id: number) => post<{ ok: true; status: 1 }>(`/api/admin/channels/${id}/disable`, {}),
  test: (id: number, model?: string) => post<TestResult>(`/api/admin/channels/${id}/test`, { model }),
  batchTest: (ids: number[], model?: string) => post<{ results: TestResult[] }>('/api/admin/channels/batch_test', { ids, model }),
  health: (id: number) => get<Channel['health']>(`/api/admin/channels/${id}/health`),
  listMappings: (id: number) => get<{ items: Mapping[] }>(`/api/admin/channels/${id}/model_mappings`),
  putMappings: (id: number, mappings: Mapping[]) => post<{ items: Mapping[] }>(`/api/admin/channels/${id}/model_mappings`, { mappings }),
}
